Store the domain genesis storage on the consensus chain

NingLin-P opened the pull request:
close #1798 

In domain v2, the domain's genesis state root is driven and stored on the consensus chain when the domain is instantiated, and we need to ensure when the operator node is started it will construct the same genesis state root. 

Currently, this is done by storing the domain runtime code and optionally the domain `GenesisRuntimeConfig` on the consensus chain, and the operator node will fetch these data from consensus chain (use `GenesisRuntimeConfig::default` in case it is not stored) and them construct the genesis state by `<GenesisRuntimeConfig as BuildStorage>::build_storage`.

Unfortunately, both `GenesisRuntimeConfig::default` and `<GenesisRuntimeConfig as BuildStorage>::build_storage` are called on the client side and the return value is determined by the native runtime that compiled with the consensus chain, thus even if we store the same WASM runtime code onchain the operator may still get a different genesis state root if the operator is running with a newer release of `subspace-node`, see #1798 for more detail.

~~This PR fixes this issue by directly storing the domain genesis storage onchain when domain is instantiated instead of building the genesis storage when the operator is started. This won't bring much cost compared with the main branch because:~~
- ~~We remove the runtime code from the genesis storage to avoid duplicating with `RuntimeRegistry`~~
- ~~We just changed to store the genesis storage instead of `GenesisRuntimeConfig` onchain~~

This PR fixes the issue by storing the domain genesis storage in the runtime registry, so when registering a runtime instead of providing just the runtime code we need to provide the full genesis storage (which includes the runtime code), and every domain instance will reuse the same genesis storage with a slight adjustment of the `SelfDomainId` storage, without the need of rebuilding the genesis storage again.

This PR also comes with a domain subcommand `build-genesis-storage` that we can use to get the raw genesis storage for registering runtime.

This PR also gets rid of the ugly workaround of constructing chain spec with the `GenesisRuntimeConfig` which is introduced in #1668 

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

vedhavyas reviewed (commented):
I'm still not sure the exact reason we need to hold entire genesis state 🤔 
If the `build_storage` relies on std, then other substrate chains might not sync at all but that is not the case. 
Can you add more detail as to what is actually causing the different state root? If there is an actual requirement for std, then the fix should be sent upstream instead of trying to patch it here IMO

vedhavyas commented on crates/sp-domains/src/lib.rs:
```diff
@@ -607,6 +609,12 @@ impl DomainsDigestItem for DigestItem {
 }
 
+/// The storage key of the `SelfDomainId` storage item in the `pallet-domain-id`
+pub const SELF_DOMAIN_ID_STORAGE_KEY: [u8; 32] = [
```
Please construct an actual storage and derive the key instead of magic bytes

NingLin-P replied:
We can construct something `SelfDomainId::<T>::hashed_key().to_vec()` in the runtime but not the client side, I also consider runtime API but this key is used in the consensus node, thus we need to add the `pallet-domain-id` to consensus runtime to do this, I'm not sure if it is worth, so I just add a unit test to `pallet-domain-id` to ensure it won't changed accidentally.

vedhavyas replied:
We already derive something similar here - https://github.com/subspace/subspace/blob/main/domains/primitives/messenger/src/messages.rs#L366 without a runtime call. Introducing this will alleviate the readability problem rather than seeing some magic bytes. Ofcourse, the storage definition on the runtime is not directly connected, hence, this client side impl would need to be updated accordingly. I doubt we will change storage definitions that often and even if they did client usage can follow the similar usage

vedhavyas commented on crates/pallet-domains/src/runtime_registry.rs:
```diff
@@ -134,19 +94,26 @@ pub(crate) fn register_runtime_at_genesis<T: Config>(
 pub(crate) fn do_schedule_runtime_upgrade<T: Config>(
     runtime_id: RuntimeId,
-    code: Vec<u8>,
+    raw_genesis_storage: Vec<u8>,
```
we already have the raw_genesis when we register a runtime. Why do we need this again ?

NingLin-P replied:
Because the genesis state will be changed as runtime version changes as observed previously, so every different runtime needs a corresponding genesis storage.

nazar-pc replied:
This doesn't seem right. If we upgrade existing domain, we shouldn't wipe the state and replace it with one from genesis. Domain should do the actual upgrade instead with provided updated code.

NingLin-P replied:
> Domain should do the actual upgrade instead with provided updated code.

Yeah, you are right!

vedhavyas replied:
Still not sure why we need to pass the entire geneis storage here. We just new runtime and the rest should stay as is. What am I missing here ?

NingLin-P replied:
> Because the genesis state will be changed as runtime version changes as observed previously, so every different runtime needs a corresponding genesis storage.

^

@nazar-pc 's comment is wrong, plz take a look at https://github.com/subspace/subspace/pull/1824#discussion_r1326576232

nazar-pc replied:
Passing genesis storage means you can instantitate domain with correct storage right away, otherwise you'll need to do multiple runtime upgrades right after instantiation, which makes little sense.

vedhavyas reviewed (commented):
I'm still not sure why we need the host function :thinking:. We already have the raw chainspec stored, when the domain instance is created, we can make necessary domain specific changes and generate the state root which is what the host function is doing now. I was hoping the host function is completely removed in this change

nazar-pc commented on crates/sp-domains/src/storage.rs:
```diff
@@ -0,0 +1,75 @@
+type GenesisStorage = BTreeMap<StorageKey, StorageData>;
+
+/// Raw storage content for genesis block, it will be serialized and stored on the consensus chain
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
```
Does it have to be JSON? Genesis is a relatively large data stucture and it will be much more efficient to work with in binary form.

nazar-pc commented on domains/pallets/domain-id/src/lib.rs:
```diff
@@ -44,16 +47,17 @@ mod pallet {
-                self.domain_id
-                    .expect("Genesis config of pallet-domain-id must be set"),
-            );
+            // NOTE: `domain_id` can set to `None` in the `RuntimeGenesisConfig` of chain spec since
```
```suggestion
            // NOTE: `domain_id` can be set to `None` in the `RuntimeGenesisConfig` of chain spec since
```

nazar-pc commented on domains/pallets/domain-id/src/lib.rs:
```diff
@@ -44,16 +47,17 @@ mod pallet {
     impl<T: Config> Pallet<T> {
         pub fn self_domain_id() -> DomainId {
             SelfDomainId::<T>::get()
+                .expect("self domain-id must set to Some when building the genesis storage")
```
I don't get it, `fn build` above works with `Option`, I don't see anything that ensures the ID is not `None`, hence I don't understand why it is fine to panic here.

NingLin-P replied:
Because the `SelfDomainId` must be reset in the genesis storage after the original genesis storage is built, the workflow:
1. The runtime is registered with the genesis storage, where the `SelfDomainId` is none
2. The domain is instantiated with a new domain id, the `SelfDomainId` of the original genesis storage is reset with the new domain id after that the genesis storage is used in the domain instance chain.

nazar-pc replied:
Hm, then the comment should be something like this:
```
Domain ID must be set during domain instantiation
```

It is neither set to `Some` (it is the key that is missing, not that we store `None`) nor we do this during building of genesis storage, we do this during domain instantiation instead. I think it is helpful to be accurate here to prevent confusion.

nazar-pc commented on test/subspace-test-client/src/domain_chain_spec.rs:
```diff
@@ -84,9 +84,6 @@ pub fn testnet_evm_genesis() -> GenesisConfig {
-            // Id of the genesis domain
-            domain_id: Some(DomainId::new(0)),
-        },
+        self_domain_id: Default::default(),
```
Why do you even need to specify it here? It seemed like it doesn't matter what it is set to anyway.

NingLin-P replied:
This field is set to `Some(DomainId::new(123))` in a later commit for testing purposes, but yeah we can use `..Default::default()` to replace `pallet_xxx: Default::default()`

nazar-pc replied:
Why is this field necessary here at all? Wouldn't it be overridden during domain instantiation anyway?

NingLin-P replied:
It is set to `Some(DomainId::new(123))` for testing if it will be overridden, in the production chain spec we can replace them with `..Default::default()`

nazar-pc replied:
If it is overridden or doesn't matter in either case, why is it there at all?

NingLin-P replied:
Sorry, I'm not following, it is a field of `RuntimeGenesisConfig` so it should be either specified with a value `pallet_domain_id: ...` or use `..Default::default()`, and it is set to `Some(DomainId::new(123))` here for testing purposes.

nazar-pc commented on crates/sp-domains/src/lib.rs:
```diff
@@ -666,10 +682,39 @@ pub trait Domain {
+                return None;
+            }
+        };
+        let hash = BlakeTwo256::hash_of(&code);
```
Why can't these be done in runtime without host function call? This looks like simple decoding of JSON and then getting values from well known keys to me unless I miss something. We don't need to work with "Storage" as such.

nazar-pc commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -810,14 +809,18 @@ mod pallet {
             origin: OriginFor<T>,
             runtime_name: Vec<u8>,
             runtime_type: RuntimeType,
```
Now I'm wondering why do we need these two fields, why would generic `pallet-domains` need to care that there exists EVM domain at all? Name also seens unnecessary on the surface.

NingLin-P replied:
`runtime_name` is not used except for display purposes (like querying chain state in polkadot.js), `runtime_type` is still necessary for starting the operator node until #1877 is resolved.

nazar-pc commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -810,14 +809,18 @@ mod pallet {
             runtime_name: Vec<u8>,
             runtime_type: RuntimeType,
-            code: Vec<u8>,
+            raw_genesis_storage: Vec<u8>,
```
From what I recall raw genesis storage was a JSON string, why is it a vector of bytes here (and in other places too) and not a string? JSON is not a binary format. I think I already commented about this in one of the other PRs.

NingLin-P replied:
No specific reason, just follow previous/upstream practice

nazar-pc replied:
I think we should change it to string everywhere we have it then because it is a string.

nazar-pc replied:
I see it is decoded from raw bytes later, can't we take the decoded instance here as an argumnet then? Otherwise we're encoding/decoding it back and forth unnecessarily. Also I just noticed `runtime_name` is a vector of bytes for some reason rather than a string too :confused:

NingLin-P replied:
Do you mean `raw_genesis_storage: RawGenesis`?  that may not be friendly for those who want to initiate this call through `polkadot.js` or something similar, with `Vec<u8>` the user can upload a file directly.

Will change `runtime_name` and `domain_name` to string in later commit.

nazar-pc replied:
Are you sure they can't upload the file directly with other data types? I think at least when implementing some trait on it it should be possible to do so.

NingLin-P replied:
I think it will treat `RawGenesis` as `struct` data type and request to provide each of its fields, and I just verified:

with `raw_genesis_storage: Vec<u8>`:
<img width="825" alt="截屏2023-09-15 06 30 39" src="https://github.com/subspace/subspace/assets/28501710/af975ab1-23ee-41ee-989c-44798a94ca0a">

with `raw_genesis_storage: RawGenesis`:
<img width="822" alt="截屏2023-09-15 06 31 06" src="https://github.com/subspace/subspace/assets/28501710/71395d54-9f57-4faf-b846-8ee3cdb07f2c">

nazar-pc replied:
Hm, I see, but I'm still wondering if we can provide some kind of metadata to make upload work there like you expect it to

NingLin-P replied:
> if we can provide some kind of metadata to make upload work there like you expect it to

I didn't find this after a few searches and gave up but a TODO was added

>  Also I just noticed runtime_name is a vector of bytes for some reason rather than a string too 😕

I tried to use `String` in runtime but `sp_std::alloc::String` is not supported, I did try `scale_info::prelude::string::String` and got it complied but it does not seem legitimate, I also found some comments in substrate that suggest we shouldn't use `String` in runtime:
https://github.com/paritytech/polkadot-sdk/blob/9e403629d590abcd57c13f4010eb0714a4ee6642/substrate/frame/contracts/primitives/src/lib.rs#L67-L68
https://github.com/paritytech/polkadot-sdk/blob/9e403629d590abcd57c13f4010eb0714a4ee6642/substrate/primitives/std/without_std.rs#L43-L45

nazar-pc replied:
I would argue we should use `sp-std` less, ideally not use at all and just write idiomatic Rust code, which in this case means just using `alloc::string::String`. So many changes during Substrate upgrades are due to re-exports as well.

nazar-pc reviewed (commented):
I will not pretend to understand what is going on, but I have questions and concerns. The goal was to remove host calls, but we seem to still have them, which is not good.

NingLin-P committed:
Refactor the pallet-domain-id and export the storage key of SelfDomainId

This storage key will be used in upcoming commit to reset the SelfDomainId storage
item to the correct value in the domain genesis storage

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Introduce RawGenesis which is the genesis storage with codec and TypeInfo supported

RawGenesis is mostly port from upstream with slight changes to support our usecase

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Store the RawGenesis in the runtime registry instead of just the runtime code

In this way the domain instance can reuse the genesis storage, which including
both the wasm runtime code and genesis config of other pallet, and avoid to build
the genesis storage again

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Use the runtime_object.raw_genesis to derive the genesis storage/state root for domain instance

This commit remove the usage of host function to derive genesis receipt and also
get rid of the ugly workaround of constructing chain spec with the RuntimeGenesisConfig
and global variable

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Remove the domain related host function

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Regiter the genesis domain on consensus block zero

Previously we need to delay this to block one due to the host function is
unavailable but now the host function is removed. Also, because the block
hash of the genesis block is unavailable thus the genesis receipt's consensus_block_hash
is set to the default value

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Remove the raw_genesis_config from domain registry

Since we use runtime_object.raw_genesis to derive the genesis state for domain
instance and which already contains the sudo account, raw_genesis_config is
needless now

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Replace the domain subcommand build-genesis-config with build-genesis-storage

Since now we don't need the raw_genesis_config when instantiate domain and
we need the raw_genesis_storage when register runtime

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Increase the spec_version of the test evm domain runtime by one

This is used to simulate the situation where the domain runtime is upgraded after
domain instantiation, thus the domain tests will ensure the domain genesis state
will always be the same even it is generated in different place with different native
runtime. This commit also change the domain_id of the RuntimeGenesisConfig to an arbitrary
value to ensure the domain_id will be reset to the correct value in the genesis storage

Signed-off-by: linning <linningde25@gmail.com>

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1158,13 +1162,9 @@ mod pallet {
-                    )
-                    .expect("Genesis domain instantiation must always succeed");
+                let domain_id =
+                    do_instantiate_domain::<T>(domain_config, domain_owner.clone(), One::one(), None)
```
we do not need the `raw_genesis_config` to instantiate domain anymore. We should also remove the sudo related things from the domain_registry as well

NingLin-P replied:
It is removed in a later commit but I'm not sure what you means by `remove the sudo related things from the domain_registry as well`

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -520,16 +523,11 @@ mod pallet {
     /// within this block are verified.
-    /// The storage is cleared on block finalization.
+    /// TODO: The storage is cleared on block finalization that means this storage is already cleared when
+    /// verifying the `submit_bundle` extrinsic and not used at all
```
this is never meant to be used outside the runtime. This storage purely exists to ensure POE is verified if there was a Epoch transition in this block

NingLin-P replied:
Yeah, but the storage is inserted during block execution and cleared on block finalization which means it only exists during block execution, and we need to verify POE whenever a node receives a `submit_bundles` extrinsic.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1493,26 +1477,19 @@ impl<T: Config> Pallet<T> {
-                    .get(operator_id)
-                    .ok_or(BundleError::BadOperator)?;
-                Ok((*operator_stake, pending_election_params.total_domain_stake))
+        if let Some(pending_election_params) = LastEpochStakingDistribution::<T>::get(domain_id) {
```
This refactor seems unrelated :thinking:

NingLin-P replied:
It is related to the change of registering the genesis domain at block `#0` instead of block `#1`.

Because when in block `#1`, the `LastEpochStakingDistribution` is cleared when verifying the bundle that is to be included at `#2` so we always use `DomainStakingSummary`.

But when in block `#0` namely the genesis block there is no block finalization thus the `LastEpochStakingDistribution` is used when verifying the bundle  that is to be included at `#1`, and it will fail with `BadOperator` due to the operator id not present in `LastEpochStakingDistribution::operators`

vedhavyas commented on crates/sp-domains/src/lib.rs:
```diff
@@ -391,13 +387,13 @@ impl<
             domain_block_extrinsic_root: Default::default(),
             parent_domain_block_receipt_hash: Default::default(),
-            consensus_block_hash: consensus_genesis_hash,
+            consensus_block_hash: Default::default(),
```
not sure why? Ideally, this should be the block hash in which the domain was instantiated. Is that not so ?

NingLin-P replied:
Sorry, there should be a TODO here, but the change is due to the genesis domain being registered at the genesis block now, and we can't know the hash of the genesis block at that time.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1014,19 +1016,12 @@ mod pallet {
-                ensure_root(origin)?;
-                (T::SudoId::get(), Some(raw_genesis))
-            };
+            let who = ensure_signed(origin)?;
```
SudoId from the config should be removed as well. Was introduced for the same reason of raw_genesis

NingLin-P replied:
Do you mean remove it from the domain runtime's `RuntimeGenesisConfig`?

vedhavyas replied:
Yeap. We dont need to any Sudo here.

NingLin-P replied:
The reason we can remove the sudo/raw_genesis from `instantiate_domain` is the new domain instance will use the runtime object's genesis state which already contains the sudo. If the sudo account is removed from the domain runtime's `RuntimeGenesisConfig` then the runtime object's genesis state (and all the domain chain) won't have the sudo at all, it can be an issue when we launch the next gemini network.

vedhavyas replied:
Sorry, I did not communicated that well. I was suggesting to remove `SudoID` config paramater

vedhavyas commented on test/subspace-test-client/src/domain_chain_spec.rs:
```diff
@@ -89,12 +88,12 @@ pub fn testnet_evm_genesis() -> RuntimeGenesisConfig {
         },
-        ethereum: Default::default(),
-        base_fee: Default::default(),
         self_domain_id: evm_domain_test_runtime::SelfDomainIdConfig {
```
why not just default instead ?

NingLin-P replied:
the `domain_id` is set to `Some(DomainId::new(123))` for testing purpose

vedhavyas reviewed (commented):
Make sense. Some nits and questions.

nazar-pc commented on crates/sp-domains/src/storage.rs:
```diff
@@ -0,0 +1,157 @@
+
+/// Create a new empty instance of in-memory backend.
+///
+/// NOTE: this function is port from `sp_state_machine::in_memory_backend::new_in_mem` which is
+/// only export for `std` but we need to use it in `no_std`
```
What about submitting PR upstream that exposes it in other environments? I'd pull it earlier today with Substrate upgrade already.

NingLin-P replied:
I want to get some feedback for this direction first, if it is the right direction I will surely contribute it upstream.

nazar-pc commented on crates/sp-domains/src/storage.rs:
```diff
@@ -0,0 +1,157 @@
+    TrieBackendBuilder::new(db, empty_trie_root::<LayoutV0<H>>()).build()
+}
+
+// NOTE: this is port from `sp_core::storage::StorageKey` with `TypeInfo` supported
```
Again, looks like something worth contributing upstream.

nazar-pc commented on crates/pallet-domains/src/runtime_registry.rs:
```diff
@@ -73,10 +93,18 @@ pub(crate) fn can_upgrade_code(
+    let mut raw_genesis: RawGenesis = Decode::decode(&mut raw_genesis_storage.as_slice())
+        .map_err(|_| Error::FailedToDecodeRawGenesis)?;
+
+    // Remove the runtime code from `raw_genesis` before storing it
```
Why removing runtime code from genesis state? It would make sense to me to leave it in there so you don't need to put it back in later.

NingLin-P replied:
Because runtime code is updatable, it is weird to extract the code, update it, and store the updated code back to the genesis state.

nazar-pc replied:
> Because runtime code is updatable, it is weird to extract the code, update it, and store the updated code back to the genesis state.

This should never be necessary, runtime code in genesis state doesn't need to be updatable. It needs to be canonical instead and come as a bundle, just like it is before you remove runtime code out of it. For updated runtime genesis state will be different too. State and code should come in pairs when it comes to instantiation.

You don't update genesis state without updating runtime code and you don't update runtime code without updating genesis state.

NingLin-P replied:
Okay, I may be misled by your reply https://github.com/subspace/subspace/pull/1824#discussion_r1320247718, that is what I thought at first! the genesis state and the runtime code should come in pairs. This should be true for both the first registration and the later runtime upgrade. The new domain instance should use both the upgraded runtime code and genesis state to initialize its genesis block, this is the only place the genesis state is involved. The existing domain instance should just upgrade its runtime code don't care about the updated genesis state.

nazar-pc replied:
The last part of above comment is correct. They come together, but during upgrades we only upgrade code and let it change the state accordingly.

NingLin-P replied:
The existing domain instance will definitely upgrade its code and change its state accordingly, what I mean is for domain instances that are instantiated after the runtime code is upgraded in the consensus chain, they should use both the upgraded code and genesis state to initialize their genesis block.

nazar-pc replied:
Hm, it depends on how it is implemented. The way I see it is when code changes it is essentially a different version of domain, so you upload both new genesis state and code and then point existing domain to the new code. What you're suggesting seems to be that domain code is updated "in place", which is confusing for other domains that use the old code (will old code be just "gone" or swapped forcefully with a new version?), but also implies that any future version of the code should be able to upgrade from any past version of the state, which at some point might take so long it wouldn't instantiate at all (potentially).

But keep in mind that I don't have latest design of domains in my head.

vedhavyas replied:
I think it makes sense to leave the Initial runtime in the Genesis state itself. Sure it helps in retrieving the domain runtime faster. But apart from that, I dont see any benefit but rather brings confusion where is a new runtime. 
I say, we store all the new runtime seperately as they are submitted but leave the first one in place. This also avoids the manually refilling them when we need the full genesis

nazar-pc replied:
It still means we need to update state one step at a time when we deploy new domain. Maybe this is acceptable, but feels inefficient.

NingLin-P replied:
I'm okay with storing the runtime code along with the genesis state, but what we need to discuss is whether the genesis state should be updated as the domain runtime upgraded in the consensus chain. My opinion is yes, the domain genesis state and runtime code should come in pairs and be updated at the same time.

> What you're suggesting seems to be that domain code is updated "in place", which is confusing for other domains that use the old code
> I say, we store all the new runtime seperately as they are submitted but leave the first one in place. This also avoids the manually refilling them when we need the full genesis

Suppose the domain is instantiated, it commits to a runtime object with the specified `RuntimeId`, and uses both the runtime object's runtime code and genesis state to initialize the domain chain.

Later, the runtime object is updated (both the code and genesis state), and the existing domain instance will upgrade its runtime code unconditionally and automatically as the upgrade is derived from the consensus block deterministically. However, the existing domain doesn't care about the updated genesis state because it only needs this when building its genesis block.

Then there is a new domain instance instantiated and committed to the same runtime object with the same `RuntimeId`, now the new domain instance should use the newest version of both the runtime code and genesis state to initialize its genesis block. This is the reason why I think we should update the runtime object's runtime code and genesis state at the same time.

And I can't see why we need to store all the version of the runtime code in the consensus chain state to support instantiating new domain instances with an older version of the runtime code.

nazar-pc replied:
> My opinion is yes, the domain genesis state and runtime code should come in pairs and be updated at the same time.

That is my expectation and preference as well.

> And I can't see why we need to store all the version of the runtime code in the consensus chain state to support instantiating new domain instances with an older version of the runtime code.

Then whatever the latest is will have to support runtime upgrades from whatever the oldest registered domain is (even if it is not running right now) because otherwise domain will not resume. I feel like this is an issue and storing all *used* versions is very desirable.

NingLin-P replied:
> Then whatever the latest is will have to support runtime upgrades from whatever the oldest registered domain is (even if it is not running right now) because otherwise domain will not resume

No, as said above the domain runtime upgrade is not controlled by the domain owner but by the runtime object, whenever the runtime object is upgraded the domain instance will upgrade its runtime unconditionally and automatically.

If the domain instance node is not running for a while, then it should catch up to the domain chain tip by deriving the domain block from the consensus block, it can't miss any consensus block since it is instantiated unless we support syncing the domain chain from other nodes, but still all the consensus block should be handled by some (but not all) domain nodes in case any potential bundle/runtime upgrade is missed.

NingLin-P replied:
The PR is updated with runtime code and genesis state stored together and updated together, PTAL.

nazar-pc reviewed (commented):
I don't think I follow everything through, but I certainly like this more than previous iteration

NingLin-P committed:
Update both the runtime code and the genesis state in the upgrade_domain_runtime call

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into genesis-storage

NingLin-P committed:
Use String for runtime_name/domain_name

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into genesis-storage

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -31,6 +31,8 @@ mod staking;
 mod staking_epoch;
 pub mod weights;
 
+extern crate alloc;
```
Dont think this is required, is it ?

NingLin-P replied:
It required by `alloc::string::String`

nazar-pc replied:
`extern crate alloc` is basically the only case where this old syntax is both the only option and required at the same time.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1379,7 +1361,7 @@ impl<T: Config> Pallet<T> {
     pub fn domain_runtime_code(domain_id: DomainId) -> Option<Vec<u8>> {
         RuntimeRegistry::<T>::get(Self::runtime_id(domain_id)?)
-            .map(|runtime_object| runtime_object.code)
+            .and_then(|mut runtime_object| runtime_object.raw_genesis.take_runtime_code())
```
This doe not consider runtime upgrades and instead return the first runtime

NingLin-P replied:
We don't maintain historical domain runtime code, they are updated in place and the returned value is determined by the `consensus_hash` used to call the runtime API.

vedhavyas replied:
No, that should not be case. We should never touch the raw chainspec and treat it just like that for domain raw spec as well. Updating in place means, raw chainspec is always updating whenever there is a runtime upgrade.

nazar-pc replied:
This is for the purposes of future instantiation such that we don't have to go through potentially many runtime upgrades and keep the whole upgrade path support in the runtime forever (though maybe we do need that for other reasons).

vedhavyas replied:
~Then we will essential break the exisiting domains for the operators who want to sync from beginning, IIUC. If we do want to support runtime upgrades such that new domains do not need to go through the upgrade path. Then each domain should ideally carry their current rutime hash and fetch the right runtime based on that hash. This way, exisiting operator domains can sync from the beginning, while new domains and their operators can start from the latest upgrade as their genesis runtime.~

Update: Okay I see where my confusion is coming from. Fetching this from chain_spec and the fact that we are updating this in place leads to this confusion. We already update the chainspec with `domain_id`, why not have a Runtime storage against runtime hash and then return that instead ?

NingLin-P replied:
Sorry, I don't get your point. `raw_genesis` is not fetched from the chain spec but built from chain spec, different runtime code can lead to different `raw_genesis` even if the rest of the chain spec is the same.

> We already update the chainspec with domain_id, why not have a Runtime storage against runtime hash and then return that instead ?

Do you mean storing the runtime code separately? the runtime code and `raw_genesis` are come together and updated together, `domain_runtime_code` is the only place that runtime code is used alone so I don't see much benefit to doing so, also see https://github.com/subspace/subspace/pull/1824#discussion_r1326332807

vedhavyas commented on crates/pallet-domains/src/runtime_registry.rs:
```diff
@@ -23,24 +25,40 @@ pub enum Error {
-    pub code: Vec<u8>,
+    // The raw gensis storage that contains the runtime code.
+    // NOTE: don't use this field directly but `into_complete_raw_genesis` instead
+    pub raw_genesis: RawGenesis,
```
then we not make it public instead ?

NingLin-P replied:
It is `pub` for the test.

vedhavyas replied:
if you expose a function, then this is not required TBH

NingLin-P replied:
It is used to construct the `RuntimeObject`

vedhavyas commented on crates/sp-domains/src/lib.rs:
```diff
@@ -21,12 +21,17 @@ pub mod bundle_producer_election;
 pub mod valued_trie_root;
 pub mod verification;
 
+extern crate alloc;
```
not sure if this is required, is it ?

NingLin-P replied:
ditto

vedhavyas commented on crates/subspace-node/src/domain/evm_chain_spec.rs:
```diff
@@ -318,8 +275,8 @@ fn testnet_genesis(
         },
-        transaction_payment: Default::default(),
         balances: BalancesConfig {
+            // TODO: remove `endowed_accounts` once XDM is ready
```
I dont understand this TODO. Why do we want to remove endowed accounts ?

NingLin-P replied:
Because the genesis storage is registered with the runtime code and is reused by every domain instance, so it basically means we mint SSC whenever a domain is instantiated.

vedhavyas replied:
sure, for the production runtime these will be empty anyway. For the tests, I would prefer having these endowed accounts rather than using the XDM to moving funds from the Consensus. This is a bad ux when testing or developing

NingLin-P replied:
Okay, will update the comment

vedhavyas reviewed (commented):
General Direction make sense but left some coments

vedhavyas reviewed (approved):
Okay. Lets move on.
Thank you for being patient
